Implement basic types

DaGuich opened the pull request:
Implement the basic data types defined by the OSM wiki. 

https://wiki.openstreetmap.org/wiki/Elements

DaGuich committed:
Implemented basic types

DaGuich committed:
Documentation of the data types

Descriptions of the OSM wiki are mainly used

DaGuich committed:
Tune up documentation

Made links to osm wiki real links

amandasaurus commented:
Did i miss something, but it looks like the relation type doesn't store the members?

The lat lon of nodes can be null, and this can happen in data files. I can't remember for sure when this happens, either when a node is deleted (common enough), or redacted (much less common).

On 22 June 2018 21:11:18 CEST, Matthias Gilch <notifications@github.com> wrote:
>Implement the basic data types defined by the OSM wiki. 
>
>https://wiki.openstreetmap.org/wiki/Elements
>
>TODO: 
>
>- Documentation
>You can view, comment on, or merge this pull request online at:
>
>  https://github.com/georust/rust-osm/pull/3
>
>-- Commit Summary --
>
>  * Implemented basic types
>
>-- File Changes --
>
>    M src/lib.rs (44)
>    A src/test.rs (3)
>
>-- Patch Links --
>
>https://github.com/georust/rust-osm/pull/3.patch
>https://github.com/georust/rust-osm/pull/3.diff

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

DaGuich committed:
Move types in a module

TeXitoi commented on src/lib.rs:
```diff
@@ -1,3 +1,66 @@
+pub struct ElementInfo {
+    pub id: i64,
+    pub user: String,
+    pub uid: i32,
```
The user fields should be derived from the uid field, no?

DaGuich replied:
Both, user and uid fields are present in the OSM-XML format. It's not explicitly specified if those fields are optional.

amandasaurus replied:
OSM users can change their username, so the user field will (I think?) be the name they used at the time. uids won't changed. See below for optional user fields

TeXitoi commented on src/lib.rs:
```diff
@@ -1,3 +1,66 @@
+    pub id: i64,
+    pub user: String,
+    pub uid: i32,
+    pub timestamp: String,
```
Better to use a dedicated object as a chrono::DateTime: more compact and feature full.

TeXitoi commented on src/lib.rs:
```diff
@@ -1,3 +1,66 @@
+/// See: [OSM wiki - Node](https://wiki.openstreetmap.org/wiki/Node)
+pub struct Node {
+    pub element_info: ElementInfo,
+    pub tags: HashMap<String, String>,
```
Tags are also common between the different types. A HashMap takes a lot of space. As there is not so much tags in an OSM object, better to use a Vec or a flat_map::FlatMap

TeXitoi commented on src/lib.rs:
```diff
@@ -1,3 +1,66 @@
+    pub element_info: ElementInfo,
+    pub tags: HashMap<String, String>,
+    pub lat: f64,
+    pub lon: f64
```
These fields can be encoded using i32 for compactness (10^-7 degrees)

DaGuich replied:
I went along with the documentation in the wiki which says: 
> Do not use IEEE 32-bit floating point data type since it is limited to about 5 decimal places for the highest longitude.

But it also says: 
> A 32-bit method used by the Rails port is to use an integer (by multiplying each coordinate in degrees by 1E7 and rounding it: this allows to cover all absolute signed coordinates in ±214.7483647 degrees, or a maximum difference of 429.4967295 degrees, a bit more than what is needed).

See: [OSM Wiki - Node](https://wiki.openstreetmap.org/wiki/Node#Structure)

TeXitoi replied:
Yes that's my i32 proposition.

TeXitoi reviewed (commented):
By experience, you want to minimize the size of the OSM objects because you tend to stock a lot of them.

DaGuich commented:
@TeXitoi Thanks for your review. I'm relatively new with Rust, but I want to learn by practice.

amandasaurus commented:
You should make uids, usernames and changeset ids optional. Some sites (e.g. download.geofabrik.de) strip those fields from their files for user privacy reasons.

frewsxcv commented:
@DaGuich Thanks for the pull request! Considering you're the first person to put some effort into contributing to this repository, I just gave you commit access. I'd recommend continuing to use pull requests for reviews, but whenever you feel comfortable, go-ahead and merge. It's early days for this project, so don't worry about it getting everything perfect.

DaGuich commented:
@frewsxcv Thank you for the invitation. I plan to work on this project if I have time for it. So please don't expect daily work on it. But I have interest do develop a usable OSM library for Rust, which can be used by your huge ecosystem.

DaGuich committed:
Compact nodes

For compactness latitude and longitudes are represented as a
32 bit signed integer

DaGuich committed:
Optional fields

The common attributes user, uid and changeset are ofter truncated by
some data vendors. So the fields are made
Optional

DaGuich committed:
Update .gitignore for VSCode

DaGuich committed:
Move Tags in ElementInfo, rustfmt, Use chrono

Tags are moved to ElementInfo because they are used for every OSM type

Used rustfmt to sources. As a cleanup

For datetime do not use String anymore but chrono::DateTime for a more
convenient usage.

DaGuich committed:
Add methods and according tests

DaGuich committed:
Remove unused test case in test.rs

DaGuich committed:
Implement From trait for Node

Add conversion from Node to geo_types::Point
